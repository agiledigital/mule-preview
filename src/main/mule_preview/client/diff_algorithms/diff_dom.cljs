(ns mule-preview.client.diff-algorithms.diff-dom
  "Uses the diffDOM Javascript library to diff data structures
   https://github.com/fiduswriter/diffDOM"
  (:require
   [clojure.walk :refer [prewalk]]
   ["diff-dom" :refer (DiffDOM)]))

(defn- node-to-dom [node]
  (if (:type node)
    (let [{:keys [tag-name content type attributes location labels]} node
          other-attributes (dissoc node :type-tag :content :type :labels :attributes :location)]
      {:nodeName tag-name
       :childNodes content
       :attributes (merge attributes other-attributes)
       :original-type type
       :original-labels labels
       :location location})
    node))

(defn dom-to-node [node]
  (if (:nodeName node)
    (let [{:keys [nodeName childNodes attributes original-type original-labels location]} node]
      (merge {:tag-name nodeName
              :content childNodes
              :type (keyword original-type)
              :labels original-labels
              :location location} attributes))
    node))

(defn mast->dom [mast]
  (prewalk node-to-dom mast))

(defn dom->mast [dom]
  (prewalk dom-to-node dom))

(defn clojurise [js-obj]
  (let [round-trip (-> js-obj js/JSON.stringify js/JSON.parse)]
    (js->clj round-trip :keywordize-keys true)))

(defn diff [a b]
  (let [dd (new DiffDOM)
        dom-a (mast->dom a)
        dom-b (mast->dom b)]
    (clojurise (.diff dd (clj->js dom-a) (clj->js dom-b)))))