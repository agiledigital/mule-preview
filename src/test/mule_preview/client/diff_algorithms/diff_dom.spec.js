import { diff } from "../../../../../dist/mule_preview.client.diff_algorithms.diff_dom";
import {
  jsToClj,
  cljToJs
} from "../../../../../dist/mule_preview.client.test_utils";
import fileAMast from "./__fixtures__/file-a.mast.json";
import fileBMast from "./__fixtures__/file-b.mast.json";

describe("when diffing two MAST data structures", () => {
  it("produces correct output", () => {
    const diffOutput = cljToJs(diff(jsToClj(fileAMast), jsToClj(fileBMast)));
    expect(diffOutput).toMatchSnapshot();
  });
});
