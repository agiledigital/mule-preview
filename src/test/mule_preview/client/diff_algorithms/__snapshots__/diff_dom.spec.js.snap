// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when diffing two MAST data structures produces correct output 1`] = `
Array [
  Object {
    "action": "modifyAttribute",
    "name": "2",
    "newValue": Object {
      "description": null,
      "hash": -83749566,
      "tag-name": "mule",
    },
    "oldValue": Object {
      "description": null,
      "hash": -83749566,
      "tag-name": "mule",
    },
    "route": Array [],
  },
  Object {
    "action": "removeElement",
    "element": Object {
      "attributes": Array [
        "horizontal",
        Object {
          "description": "example:/expunge-convolvers",
          "hash": 1066677628,
          "tag-name": "sub-flow",
        },
      ],
      "childNodes": Array [
        Object {
          "attributes": Array [
            Object {
              "description": "Flush the convolver prisms",
              "hash": 579843109,
              "tag-name": "db:no-operation-selected",
            },
          ],
          "childNodes": null,
          "location": Object {
            "column": 104,
            "line": 37,
          },
          "nodeName": "db:no-operation-selected",
          "original-labels": null,
          "original-type": "component",
        },
        Object {
          "attributes": Array [
            "vertical",
            Object {
              "description": "Did the flush occur?",
              "hash": -1105159260,
              "tag-name": "choice",
            },
          ],
          "childNodes": Array [
            Object {
              "attributes": Array [
                "horizontal",
                Object {
                  "description": null,
                  "hash": 2014296490,
                  "tag-name": "when",
                },
              ],
              "childNodes": Array [
                Object {
                  "attributes": Array [
                    Object {
                      "description": "Fetch more prisms",
                      "hash": 881435866,
                      "tag-name": "s3:no-operation-selected",
                    },
                  ],
                  "childNodes": null,
                  "location": Object {
                    "column": 86,
                    "line": 40,
                  },
                  "nodeName": "s3:no-operation-selected",
                  "original-labels": null,
                  "original-type": "component",
                },
                Object {
                  "attributes": Array [
                    Object {
                      "description": "Log number of fanglehockeys",
                      "hash": 543925965,
                      "tag-name": "logger",
                    },
                  ],
                  "childNodes": null,
                  "location": Object {
                    "column": 77,
                    "line": 41,
                  },
                  "nodeName": "logger",
                  "original-labels": null,
                  "original-type": "component",
                },
              ],
              "location": Object {
                "column": 32,
                "line": 39,
              },
              "nodeName": "when",
              "original-labels": null,
              "original-type": "container",
            },
            Object {
              "attributes": Array [
                "horizontal",
                Object {
                  "description": null,
                  "hash": -15128758,
                  "tag-name": "otherwise",
                },
              ],
              "childNodes": Array [
                Object {
                  "attributes": Array [
                    Object {
                      "description": "Log Error",
                      "hash": -818462707,
                      "tag-name": "logger",
                    },
                  ],
                  "childNodes": null,
                  "location": Object {
                    "column": 73,
                    "line": 44,
                  },
                  "nodeName": "logger",
                  "original-labels": null,
                  "original-type": "component",
                },
              ],
              "location": Object {
                "column": 23,
                "line": 43,
              },
              "nodeName": "otherwise",
              "original-labels": null,
              "original-type": "container",
            },
          ],
          "location": Object {
            "column": 48,
            "line": 38,
          },
          "nodeName": "choice",
          "original-labels": null,
          "original-type": "container",
        },
      ],
      "location": Object {
        "column": 49,
        "line": 36,
      },
      "nodeName": "sub-flow",
      "original-labels": null,
      "original-type": "container",
    },
    "route": Array [
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "example:/http-incoming-flow",
      "tag-name": "flow",
    },
    "oldValue": Object {
      "description": "example:/http-incoming-flow",
      "tag-name": "flow",
    },
    "route": Array [
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": 1816956808,
      "tag-name": "sub-flow",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": 1816956808,
      "tag-name": "sub-flow",
    },
    "route": Array [
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "",
      "tag-name": "psuedo",
    },
    "oldValue": Object {
      "description": "",
      "tag-name": "psuedo",
    },
    "route": Array [
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "",
      "tag-name": "psuedo",
    },
    "oldValue": Object {
      "description": "",
      "tag-name": "psuedo",
    },
    "route": Array [
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "Clean carburetors",
      "hash": -688832045,
      "tag-name": "scatter-gather",
    },
    "oldValue": Object {
      "description": "Clean carburetors",
      "hash": -688832045,
      "tag-name": "scatter-gather",
    },
    "route": Array [
      1,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Start Internal Noodle Engine",
      "hash": -1986262879,
      "tag-name": "http:request",
    },
    "oldValue": Object {
      "description": "Start Internal Noodle Engine",
      "hash": -1986262879,
      "tag-name": "http:request",
    },
    "route": Array [
      1,
      1,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "Overwrite the payload",
          "hash": 266487046,
          "tag-name": "set-payload",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 72,
        "line": 14,
      },
      "nodeName": "set-payload",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      0,
    ],
  },
  Object {
    "action": "removeElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "Filter recalcitrant ions",
          "hash": -2096291590,
          "tag-name": "expression-filter",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 78,
        "line": 18,
      },
      "nodeName": "expression-filter",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "Catch Exception Strategy",
      "hash": 497298084,
      "tag-name": "catch-exception-strategy",
    },
    "oldValue": Object {
      "description": "Catch Exception Strategy",
      "hash": 497298084,
      "tag-name": "catch-exception-strategy",
    },
    "route": Array [
      0,
      1,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "Processor Chain",
      "hash": 1690317879,
      "tag-name": "processor-chain",
    },
    "oldValue": Object {
      "description": "Processor Chain",
      "hash": 1690317879,
      "tag-name": "processor-chain",
    },
    "route": Array [
      1,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": null,
      "hash": -15128758,
      "tag-name": "processor-chain",
    },
    "oldValue": Object {
      "description": null,
      "hash": -15128758,
      "tag-name": "processor-chain",
    },
    "route": Array [
      1,
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Clean Carburetor 3",
      "hash": 598254326,
      "tag-name": "http:request",
    },
    "oldValue": Object {
      "description": "Clean Carburetor 3",
      "hash": 598254326,
      "tag-name": "http:request",
    },
    "route": Array [
      1,
      0,
      2,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Overwrite the payload",
      "hash": 266487046,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Overwrite the payload",
      "hash": 266487046,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "HTTP",
      "hash": 2077519509,
      "tag-name": "http:listener",
    },
    "oldValue": Object {
      "description": "HTTP",
      "hash": 2077519509,
      "tag-name": "http:listener",
    },
    "route": Array [
      0,
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Set Z level to 7000",
      "hash": 641237277,
      "tag-name": "set-variable",
    },
    "oldValue": Object {
      "description": "Set Z level to 9000",
      "hash": 1907014272,
      "tag-name": "set-variable",
    },
    "route": Array [
      0,
      0,
      2,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Overwrite the payload",
      "hash": 266487046,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Overwrite the payload",
      "hash": 266487046,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      0,
      3,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/expunge-convolvers",
      "hash": 771211107,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/expunge-convolvers",
      "hash": 771211107,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      4,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "1",
    "newValue": Object {
      "description": "For Each",
      "hash": 242079301,
      "tag-name": "foreach",
    },
    "oldValue": Object {
      "description": "For Each",
      "hash": 242079301,
      "tag-name": "foreach",
    },
    "route": Array [
      0,
      0,
      5,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Server Noodles",
      "hash": -789013297,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Server Noodles",
      "hash": -789013297,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      0,
      6,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Deduplicate",
      "hash": 251600214,
      "tag-name": "dw:transform-message",
    },
    "oldValue": Object {
      "description": "Deduplicate",
      "hash": 385704797,
      "tag-name": "dw:transform-message",
    },
    "route": Array [
      0,
      0,
      7,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "Server Noodles 2",
          "hash": 798826535,
          "tag-name": "set-payload",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 50,
        "line": 35,
      },
      "nodeName": "set-payload",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      8,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "Nothing to see here",
          "hash": 54737936,
          "tag-name": "set-payload",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 57,
        "line": 37,
      },
      "nodeName": "set-payload",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      1,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Clean Carburetor 1",
      "hash": 1112970330,
      "tag-name": "http:request",
    },
    "oldValue": Object {
      "description": "Clean Carburetor 1",
      "hash": 1112970330,
      "tag-name": "http:request",
    },
    "route": Array [
      1,
      0,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "XML to DOM",
      "hash": -255883650,
      "tag-name": "mulexml:xml-to-dom-transformer",
    },
    "oldValue": Object {
      "description": "XML to DOM",
      "hash": -255883650,
      "tag-name": "mulexml:xml-to-dom-transformer",
    },
    "route": Array [
      1,
      0,
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Oops this has a really long name",
      "hash": -1901883164,
      "tag-name": "http:request",
    },
    "oldValue": Object {
      "description": "Oops this has a really long name",
      "hash": -1901883164,
      "tag-name": "http:request",
    },
    "route": Array [
      1,
      0,
      1,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Set Payload",
      "hash": 1079074260,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Set Payload",
      "hash": 1079074260,
      "tag-name": "set-payload",
    },
    "route": Array [
      1,
      0,
      1,
      1,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "example:/strain-overflow-noodle-requests",
          "hash": -2070961420,
          "tag-name": "flow-ref",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 123,
        "line": 20,
      },
      "nodeName": "flow-ref",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      0,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "example:/strain-overflow-noodle-requests",
          "hash": -2070961420,
          "tag-name": "flow-ref",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 123,
        "line": 21,
      },
      "nodeName": "flow-ref",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      1,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "example:/strain-overflow-noodle-requests",
          "hash": -2070961420,
          "tag-name": "flow-ref",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 123,
        "line": 22,
      },
      "nodeName": "flow-ref",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      2,
    ],
  },
  Object {
    "action": "addElement",
    "element": Object {
      "attributes": Array [
        Object {
          "description": "example:/strain-overflow-noodle-requests",
          "hash": -2070961420,
          "tag-name": "flow-ref",
        },
      ],
      "childNodes": null,
      "location": Object {
        "column": 123,
        "line": 23,
      },
      "nodeName": "flow-ref",
      "original-labels": null,
      "original-type": "component",
    },
    "route": Array [
      0,
      0,
      5,
      3,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Server Noodles 2",
      "hash": 798826535,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Server Noodles 2",
      "hash": 798826535,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      0,
      8,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Nothing to see here",
      "hash": 54737936,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Nothing to see here",
      "hash": 54737936,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      1,
      0,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Overwrite the payload",
      "hash": 266487046,
      "tag-name": "set-payload",
    },
    "oldValue": Object {
      "description": "Nothing to see here",
      "hash": 54737936,
      "tag-name": "set-payload",
    },
    "route": Array [
      0,
      1,
      0,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "Nor here",
      "hash": 614804477,
      "tag-name": "message-properties-transformer",
    },
    "oldValue": Object {
      "description": "Nor here",
      "hash": 614804477,
      "tag-name": "message-properties-transformer",
    },
    "route": Array [
      0,
      1,
      0,
      2,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      5,
      0,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      5,
      1,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      5,
      2,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      5,
      3,
    ],
  },
  Object {
    "action": "modifyAttribute",
    "name": "0",
    "newValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "oldValue": Object {
      "description": "example:/strain-overflow-noodle-requests",
      "hash": -2070961420,
      "tag-name": "flow-ref",
    },
    "route": Array [
      0,
      0,
      5,
      4,
    ],
  },
]
`;
